File: server_python/engine/prediction_engine.py
class PredictionEngine:
    def __init__(self):
        self.rules = {
            "thresholds": {"outbreak": 60, "warning": 40},
            "weather_factors": {
                "blast": {"temp_min": 22, "temp_max": 30, "humidity_min": 85},
                "brown_spot": {"temp_min": 25, "temp_max": 35, "humidity_min": 80}
            }
        }

    async def predict(self, history=None, sensor_data=None, forecast=None, crop_stage="tillering"):
        history = history or []
        sensor_data = sensor_data or {}
        forecast = forecast or {}

        reasons = []
        score7d = 0
        score14d = 0

        trend = self.calculate_trend(history)
        if trend > 0:
            trend_contribution = min(trend * 20, 40)
            score7d += trend_contribution
            score14d += trend_contribution * 1.5
            reasons.append(f"Infection severity increased by {round(trend * 100)}% over the last week.")

        env_risk = self.evaluate_env_suitability(sensor_data, forecast)
        if env_risk > 0:
            score7d += env_risk * 0.4
            score14d += env_risk * 0.6
            if env_risk > 20:
                reasons.append("Upcoming weather (high humidity/temp) is highly conducive for fungal spread.")

        if sensor_data.get("moisture_current", 100) < 30:
            score7d += 15
            score14d += 20
            reasons.append("Low soil moisture identified; plants are currently under stress and vulnerable.")

        stage_multipliers = {"seedling": 1.2, "tillering": 1.5, "flowering": 1.3, "maturity": 1.0}
        multiplier = stage_multipliers.get(crop_stage, 1.0)

        score7d *= multiplier
        score14d *= multiplier

        prob7d = min(round(score7d), 100)
        prob14d = min(round(score14d), 100)

        return {
            "prob_7d": prob7d,
            "prob_14d": prob14d,
            "confidence": 0.85 if len(history) > 5 else 0.60,
            "reasons": reasons[:3],
            "level": self.determine_level(prob7d)
        }

    def calculate_trend(self, history):
        if len(history) < 2: return 0
        
        def safe_date(ts):
            from datetime import datetime
            if isinstance(ts, str):
                return datetime.fromisoformat(ts.replace('Z', '+00:00'))
            return ts
            
        sorted_history = sorted(history, key=lambda x: safe_date(x["timestamp"]), reverse=True)
        current = sorted_history[0].get("severity_score", 0)
        previous = sorted_history[1].get("severity_score", 0)
        
        if previous == 0: return 0.5 if current > 0 else 0
        return (current - previous) / previous

    def evaluate_env_suitability(self, sensor, forecast):
        risk = 0
        humidity_avg = forecast.get("humidity_avg", 0)
        temp_avg = forecast.get("temp_avg", 0)
        rain_prob = forecast.get("rain_prob", 0)

        if humidity_avg > 85: risk += 30
        elif humidity_avg > 75: risk += 15

        if 24 <= temp_avg <= 30: risk += 20
        if rain_prob > 60: risk += 10

        return risk

    def determine_level(self, prob):
        if prob >= 75: return "CRITICAL"
        if prob >= 50: return "HIGH"
        if prob >= 25: return "MEDIUM"
        return "LOW"

prediction_engine = PredictionEngine()

File: server/engine/prediction_engine.js
/**
 * Prediction Engine for Paddy Disease Management
 * Forecasts outbreak probabilities for 7d and 14d horizons.
 */

const rules = {
    thresholds: {
        outbreak: 60, // Risk score above this is an "outbreak"
        warning: 40
    },
    weather_factors: {
        blast: { temp_min: 22, temp_max: 30, humidity_min: 85 },
        brown_spot: { temp_min: 25, temp_max: 35, humidity_min: 80 }
    }
};

class PredictionEngine {
    constructor() { }

    /**
     * Analyze a specific zone to predict future risk.
     * @param {Object} history - Array of {timestamp, severity_score}
     * @param {Object} sensorData - {humidity_avg_3d, temp_avg_3d, moisture_current}
     * @param {Object} forecast - {rain_prob, humidity_avg, temp_avg} (next 72h)
     */
    async predict(history = [], sensorData = {}, forecast = {}, cropStage = 'tillering') {
        const reasons = [];
        let score7d = 0;
        let score14d = 0;

        // 1. Feature: Severity Trend (Growth Velocity)
        const trend = this.calculateTrend(history);
        if (trend > 0) {
            const trendContribution = Math.min(trend * 20, 40); // Max 40% from trend
            score7d += trendContribution;
            score14d += trendContribution * 1.5;
            reasons.push(`Infection severity increased by ${Math.round(trend * 100)}% over the last week.`);
        }

        // 2. Feature: Environmental Suitability (Forecast + Sensor history)
        const envRisk = this.evaluateEnvSuitability(sensorData, forecast);
        if (envRisk > 0) {
            score7d += envRisk * 0.4;
            score14d += envRisk * 0.6;
            if (envRisk > 20) {
                reasons.push("Upcoming weather (high humidity/temp) is highly conducive for fungal spread.");
            }
        }

        // 3. Feature: Plant Stress (Sensors)
        if (sensorData.moisture_current < 30) {
            score7d += 15;
            score14d += 20;
            reasons.push("Low soil moisture identified; plants are currently under stress and vulnerable.");
        }

        // 4. Feature: Crop Stage Multiplier
        const stageMultipliers = { seedling: 1.2, tillering: 1.5, flowering: 1.3, maturity: 1.0 };
        const multiplier = stageMultipliers[cropStage] || 1.0;

        score7d *= multiplier;
        score14d *= multiplier;

        // Cap at 100
        const prob7d = Math.min(Math.round(score7d), 100);
        const prob14d = Math.min(Math.round(score14d), 100);

        return {
            prob_7d: prob7d,
            prob_14d: prob14d,
            confidence: history.length > 5 ? 0.85 : 0.60,
            reasons: reasons.slice(0, 3), // Top 3
            level: this.determineLevel(prob7d)
        };
    }

    calculateTrend(history) {
        if (history.length < 2) return 0;
        const sorted = [...history].sort((a, b) => new Date(b.timestamp) - new Date(a.timestamp));
        const current = sorted[0].severity_score;
        const previous = sorted[1].severity_score;
        if (previous === 0) return current > 0 ? 0.5 : 0;
        return (current - previous) / previous;
    }

    evaluateEnvSuitability(sensor, forecast) {
        let risk = 0;
        // High humidity hours (simulated by avg)
        if (forecast.humidity_avg > 85) risk += 30;
        else if (forecast.humidity_avg > 75) risk += 15;

        // Temp window
        if (forecast.temp_avg >= 24 && forecast.temp_avg <= 30) risk += 20;

        // Rain
        if (forecast.rain_prob > 60) risk += 10;

        return risk;
    }

    determineLevel(prob) {
        if (prob >= 75) return 'CRITICAL';
        if (prob >= 50) return 'HIGH';
        if (prob >= 25) return 'MEDIUM';
        return 'LOW';
    }
}

module.exports = new PredictionEngine();
